### PR TITLE
[Feature] Support interface value converted to parquet

### PR DESCRIPTION
**Pull request by wenrui-meng-ai**
### Description

- Support the interface type in schema generation. It only replies on the reflect.Type which doesn't have enough information for the interface type. So I added the reflect.Value into the schema generation 

- Fix the definition level issue which doesn't have correct repetitionType tag set in the schema 
- Centralize the pointer processing which is scattered in original solution. Instead of processing it everywhere, just append the pointer or interface into stack let the dedicated function to handle it
- Int and uint is not distinguished in the original implementation where it treat all value as int. While when there is uint value it will panic. 

[Log Analyzer Framework](https://docs.google.com/document/d/1dZGKIlg3sqaNgDupLJ1tiroQvEoLuZXxg81GFJ16vYI/edit)

### How this code was tested

unit test.
`make test`

I also updated the proto_write.go to include the interface and nested interface array. 

```
cd example
go build proto_write.go
./proto_write
parquet-tools show output/proto_message.parquet
```

image

### CODE REVIEW

File: writer/proto.go [hunk 1]
```diff
@@ -10,7 +10,7 @@ import (
 
 // Pass the obj as the go struct object
 func NewParquetWriterFromProto(pFile source.ParquetFile, obj interface{}, np int64) (*ParquetWriter, error) {
-	schemaHandler, err := schema.NewSchemaHandlerFromProtoStruct(obj)
+	schemaHandler, err := schema.NewSchemaHandlerFromStruct(obj, false)
```
Comment (yisheng-oai) [inline writer/proto.go:13]: did you mean to swap this out from `NewSchemaHandlerFromProtoStruct`?

File: schema/proto.go [hunk 2]
```diff
@@ -174,11 +175,16 @@ func GetDefinedTypeTag(typ reflect.Type) (tags map[string]string, err error) {
...
-		tags, err = GetDefinedTypeTag(typ.Elem())
+		tags, err = GetDefinedTypeTag(typ.Elem(), value.Elem())
+		// only pointer requires the Optional,
+		// other types such as map, slice relies on repeated type to be considered as optional
+		tags[ParquetRepetitionType] = RepetitionTypeOptional
 	case reflect.Array, reflect.Slice:
 		tags, err = getListTag(typ)
 	case reflect.Map:
 		tags, err = getMapTag(typ)
+	case reflect.Interface:
```
Comment (yisheng-oai) [inline schema/proto.go:186]: The crux of the PR is there correct?

File: schema/schemahandler.go [hunk 3]
```diff
@@ -241,8 +241,9 @@ func (sh *SchemaHandler) GetRootExName() string {
 }
 
 type Item struct {
-	GoType reflect.Type
-	Info   *common.Tag
+	GoType  reflect.Type
+	Info    *common.Tag
+	GoValue reflect.Value
```
Comment (yisheng-oai) [inline schema/schemahandler.go:245]: nit: put this below `GoType`

**Review by yisheng-ong-ai — APPROVED** [review]
Approving to unblock. Honestly only understand this PR at a high level. Hit me up if you think it would be valuable for me to have context on this so we can do a walk though irl.

**Comment by wenrui-meng-ai** [comment]
> Approving to unblock. Honestly only understand this PR at a high level. Hit me up if you think it would be valuable for me to have context on this so we can do a walk though irl.

@ongkong Thanks for review. I think it's OK to just review it high level. There are lots of details about how parquet file organize data and encoding data. I will make sure the testing cover the use cases we need.

File: schema/proto.go (hunk 2 above)
Comment (wenrui-meng-ai) [reply schema/proto.go:186]: Yes.

File: writer/proto.go (hunk 1 above)
Comment (wenrui-meng-ai) [reply writer/proto.go:13]: Yes. I want to just keep one code path since original implementation is actually incomplete.

**Pull request merged**